packages/rollup-core/test/app/l1-to-l2-transaction-batch-processor.spec.ts

```diff
@@ -0,0 +1,372 @@
+    )
+
+    const txHash2 = getHashFromString('derp derp derp2')
+    calldataParseMap.set(calldata2, [l1ToL2Tx2])
+
+    mockedLogsProvider.transactionsByHash.set(
+      txHash2,
+      getTransactionResponse(timestamp, calldata2, txHash2)
+    )
+
+    mockedLogsProvider.logsToReturn.push(
+      getLog([singleTxTopic], singleTxContractAddress, txHash2, 2)
+    )
+
+    await l1ToL2TransactionBatchProcessor.handle(
+      getBlock(timestamp, blockNumber)
+    )
+
+    await sleep(1_000)
+
+    listener.receivedTransactionBatches.length.should.equal(
+      1,
+      'Should have received a transaction batch'
+    )
+    listener.receivedTransactionBatches[0].timestamp.should.equal(
+      timestamp,
+      'Timestamp mismatch'
+    )
+    listener.receivedTransactionBatches[0].blockNumber.should.equal(
+      blockNumber,
+      'Block number mismatch'
+    )
+    listener.receivedTransactionBatches[0].transactions.length.should.equal(
+      2,
+      'Num transactions mismatch'
+    )
+
+    listener.receivedTransactionBatches[0].transactions[0].should.eq(
+      l1ToL2Tx,
+      'tx mismatch'
+    )
+    listener.receivedTransactionBatches[0].transactions[1].should.eq(
+      l1ToL2Tx2,
+      'tx2 mismatch'
+    )
+  })
+
+  it('should handle block with multiple logs on differnt topics properly', async () => {
+    const timestamp = 123
+    const blockNumber = 0
+    const txHash = getHashFromString('derp derp derp')
+    calldataParseMap.set(calldata, [l1ToL2Tx])
+
+    mockedLogsProvider.transactionsByHash.set(
+      txHash,
+      getTransactionResponse(timestamp, calldata, txHash)
+    )
+
+    mockedLogsProvider.logsToReturn.push(
+      getLog([singleTxTopic], singleTxContractAddress, txHash, 1)
+    )
+
+    const txHash2 = getHashFromString('derp derp derp2')
+    calldataParseMap.set(calldata2, [l1ToL2Tx2])
+
+    mockedLogsProvider.transactionsByHash.set(
+      txHash2,
+      getTransactionResponse(timestamp, calldata2, txHash2)
+    )
+
+    mockedLogsProvider.logsToReturn.push(
+      getLog([batchTxTopic], batchTxContractAddress, txHash2, 2)
+    )
+
+    await l1ToL2TransactionBatchProcessor.handle(
+      getBlock(timestamp, blockNumber)
+    )
+
+    await sleep(1_000)
+
+    listener.receivedTransactionBatches.length.should.equal(
+      1,
+      'Should have received a transaction batch'
+    )
+    listener.receivedTransactionBatches[0].timestamp.should.equal(
+      timestamp,
+      'Timestamp mismatch'
+    )
+    listener.receivedTransactionBatches[0].blockNumber.should.equal(
+      blockNumber,
+      'Block number mismatch'
+    )
+    listener.receivedTransactionBatches[0].transactions.length.should.equal(
+      2,
+      'Num transactions mismatch'
+    )
+
+    listener.receivedTransactionBatches[0].transactions[0].should.eq(
+      l1ToL2Tx,
+      'tx mismatch'
+    )
+    listener.receivedTransactionBatches[0].transactions[1].should.eq(
```

Comment:
So why would we need two different batches for two different queues? As long as they have the same [block] timestamp, shouldn't they be able to be processed together? If not and one type needs to be prioritized above another type, we can handle that in the sorting logic but the downstream result of two separate batches (if the timestamps are the same) should be the same if 1 batch vs 2.

Reply:
Ah so there are two instances where the L1->L2 queue & safety queue txs are good to split up into different batches. 

1. If we batched them together we would be forced to implement a lock in Geth / batch ingestion endpoint. This is something we will likely want eventually (in order to allow for multi-tx safety queue batches) but at the moment it's not strictly required. That's because currently we only allows 1 tx to be submitted at a time to the safety queue. This means that it's not a problem if we ingest this sequence of transactions: 1) ingest an L1->L2 tx into Geth, 2) ingest a normal "sequencer" tx into Geth, 3) ingest a safety queue tx. Once we start batching txs from the safety queue or L1->L2 queue we need to reject the "sequencer" tx in step #2 in order to ensure the batch is added atomically.
2. Inside of the `BatchSubmitterService` we will need to do separate processing for L1->L2 txs & saftey queue txs. Because each originate from a different queue, a different sort of Ethereum transaction must be generated in order to add those transactions to the `CanonicalTxChain`. The data structure I see looks something like this: `some_random_batch=[[L1->L2 tx], [a bunch of sequencer txs], [safety queue tx], [some more sequencer txs], [another L1->L2 tx perhaps], ... [then eventually once we add the feature this can be a batch of multiple safety queue txs]]`